### PR TITLE
Adding callable summary_method support and enhancements to initiate_chats

### LINKED ISSUE #1595: [Issue]: Allow repetitive agents in multiple sequential chats
### Describe the issue

Currently, we are enforcing the recipient agents in multiple sequential chats to be different: https://github.com/microsoft/autogen/blob/main/autogen/agentchat/conversable_agent.py#L905

This requirement could be relaxed (but need to be careful about how to manage history where there are repetitive agents)

### Steps to reproduce

_No response_

### Screenshots and logs

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by qingyun-wu**
## Why are these changes needed?

1. Support callable summary_method to enable users to define their own methods for generating summaries.
2. Allow repetitive recipient agents in sequential chats in `initiate_chats` (closing https://github.com/microsoft/autogen/issues/1595)
3. Enhance the default behavior of "summary_method". The default is "last_msg". When set to None, it now returns an empty string as a summary (this change was implemented in this PR). This adjustment accommodates scenarios where a summary is not needed.

 
## Related issue number

Closes #1595 

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

File: autogen/agentchat/conversable_agent.py [hunk 1]
```diff
@@ -766,7 +767,7 @@ def initiate_chat(
             cache (Cache or None): the cache client to be used for this conversation.
             **context: any context information. It has the following reserved fields:
                 "message": a str of message. Needs to be provided. Otherwise, input() will be called to get the initial message.
-                "summary_method": a string specify the method to get a summary from the chat.
+                "summary_method": a string specifying the method to get a summary from the chat. Default is DEFAULT_summary_method, i.e., "last_msg".
```
Comment (sonichi) [inline autogen/agentchat/conversable_agent.py:770]: str or callable?

File: autogen/agentchat/conversable_agent.py [hunk 2]
```diff
@@ -877,13 +888,17 @@ def _summarize_chat(
...
+        elif summary_method == "last_msg" or summary_method is None:
             try:
                 summary = agent.last_message(self)["content"].replace("TERMINATE", "")
             except (IndexError, AttributeError) as e:
                 warnings.warn(f"Cannot extract summary using last_msg: {e}", UserWarning)
+        elif isinstance(summary_method, Callable):
+            summary = summary_method(recipient, self)
         else:
-            warnings.warn(f"Unsupported summary method: {method}", UserWarning)
+            warnings.warn(
```
Comment (sonichi) [inline autogen/agentchat/conversable_agent.py:899]: Shall we raise here? And add test coverage.

File: autogen/agentchat/conversable_agent.py [hunk 3]
```diff
@@ -851,24 +852,34 @@ async def a_initiate_chat(
...
         prompt: Optional[str] = None,
         cache: Optional[Cache] = None,
     ) -> str:
         """Get a chat summary from an agent participating in a chat.
 
         Args:
-            method (str): the method to get the summary.
-            agent: the participating agent in a chat.
+            summary_method (str or callable): the summary_method to get the summary.
+                The callable summary_method should take the recipient and sender agent in a chat as input and return a string of summary. E.g,
```
Comment (afourney) [inline autogen/agentchat/conversable_agent.py:864]: What happens if this is used in the context of a group chat where the messages are in the GroupChat messages structure?

**Comment by afourney** [comment]
I like this PR. A callable was the original mechanism in SocietyOfMind, and I've found it to be very useful.

**Comment by sonichi** [comment]
@ekzhu PTAL https://github.com/microsoft/autogen/actions/runs/7863956086/job/21455151948?pr=1628

File: autogen/agentchat/conversable_agent.py (hunk 2 above)
Comment (qingyun-wu) [reply autogen/agentchat/conversable_agent.py:907]: You mean raise an error?

Comment (sonichi) [reply autogen/agentchat/conversable_agent.py:907]: yes

File: autogen/agentchat/conversable_agent.py (hunk 3 above)
Comment (qingyun-wu) [reply autogen/agentchat/conversable_agent.py:872]: The "reflection_with_llm" option is using llm to summarize based on [`chat_messages_for_summary`](https://github.com/microsoft/autogen/blob/main/autogen/agentchat/groupchat.py#L481C9-L481C34) from group chat. 

The added callable takes the sender and recipient as inputs (both could be group chat manager), and the user could decide how to use info from these two agents.

File: autogen/agentchat/conversable_agent.py (hunk 2 above)
Comment (qingyun-wu) [reply autogen/agentchat/conversable_agent.py:907]: You mean raise an error?

**Comment by ekzhu** [comment]
@sonichi @qingyun-wu I pushed a fix in this branch. The test error is due to flaky pip on GitHub's Mac OS X action server.

File: test/agentchat/test_function_call.py [hunk 4]
```diff
@@ -256,10 +256,36 @@ def test_update_function():
...
+            recipient=user_proxy,
+            message="What functions do you know about in the context of this conversation? End your response with 'TERMINATE'.",
+            summary_method="reflection_with_llm",
+        )
+    except AssertionError as e:
+        print(e)
+        assert (
+            str(e)
+            == "llm client must be set in either the recipient or sender when summary_method is reflection_with_llm."
+        )
```
Comment (sonichi) [inline test/agentchat/test_function_call.py:283]: We use `pytest.raise` for this kind of test. Please check other tests.

**Review by sonichi — APPROVED** [review]

**Pull request merged**